### PR TITLE
Custom modules

### PR DESCRIPTION
**Pull request by Joao-L-S-Almeida**
Get it back.

### CODE REVIEW

File: terratorch/models/backbones/prithvi_mae.py [hunk 1]
```diff
@@ -151,13 +151,15 @@ def __init__(
 
         self.proj = nn.Conv3d(in_chans, embed_dim, kernel_size=patch_size, stride=patch_size, bias=bias)
         self.norm = norm_layer(embed_dim) if norm_layer else nn.Identity()
+        self.warn_counter = 0
```
Comment (blumenstiel) [inline terratorch/models/backbones/prithvi_mae.py:154]: Would look better as a bool. something like `self.log_warning = True`.

File: terratorch/models/backbones/vit_encoder_decoder.py
[the remark this answers is not in the document]
Comment (Joao-L-S-Almeida) [reply terratorch/models/backbones/vit_encoder_decoder.py:1]: I'll remove it. 
The scalar regression is not used.

File: terratorch/models/backbones/prithvi_mae.py (hunk 1 above)
Comment (Joao-L-S-Almeida) [reply terratorch/models/backbones/prithvi_mae.py:154]: Done.,

File: terratorch/tasks/regression_tasks.py [hunk 2]
```diff
@@ -1,7 +1,7 @@
 """This module contains the regression task and its auxiliary classes."""
 
 from collections.abc import Sequence
-from typing import Any
+from typing import Any, Optional
```
[the remark this answers is not in the document]
Comment (Joao-L-S-Almeida) [reply terratorch/tasks/regression_tasks.py]: I agree.

File: terratorch/tasks/regression_tasks.py [hunk 3]
```diff
@@ -185,22 +185,45 @@ def __init__(
...
         self.val_loss_handler = LossHandler(self.val_metrics.prefix)
         self.monitor = f"{self.val_metrics.prefix}loss"
         self.plot_on_val = int(plot_on_val)
 
+    @property
+    def model_module(self):
+        return self._model_module
+
+    def _bypass_build(self):
+        return self.model_module
```
[the remark this answers is not in the document]
Comment (Joao-L-S-Almeida) [reply terratorch/tasks/regression_tasks.py:211]: Lightning bureaucracy.

File: terratorch/models/backbones/vit_encoder_decoder.py [hunk 4]
```diff
@@ -374,13 +391,29 @@ def _init_weights(self, m):
             nn.init.constant_(m.bias, 0)
             nn.init.constant_(m.weight, 1.0)
 
+    def _pad_images(self, imgs: Tensor, patch_size:int=None, padding:str='constant') -> Tensor:
```
[the remark this answers is not in the document]
Comment (Joao-L-S-Almeida) [reply terratorch/models/backbones/vit_encoder_decoder.py:394]: I'll remove this class.

File: terratorch/tasks/regression_tasks.py [hunk 5]
```diff
@@ -398,5 +425,70 @@ def model_forward(x):
...
+                     optimizer=optimizer,
+                     optimizer_hparams=optimizer_hparams,
+                     scheduler=scheduler,
+                     scheduler_hparams=scheduler_hparams,
+                     freeze_backbone=freeze_backbone,
+                     freeze_decoder=freeze_decoder,
+                     plot_on_val=plot_on_val,
+                     tiled_inference_parameters=tiled_inference_parameters,)
+
+    def validation_step(self, batch: Any, batch_idx: int, dataloader_idx: int = 0) -> None:
```
[the remark this answers is not in the document]
Comment (Joao-L-S-Almeida) [reply terratorch/tasks/regression_tasks.py]: This part was totally removed.

File: terratorch/tasks/regression_tasks.py [hunk 6]
```diff
@@ -398,5 +425,70 @@ def model_forward(x):
...
+            batch_idx: Integer displaying index of this batch.
+            dataloader_idx: Index of the current dataloader.
+        """
+        x = batch["image"]
+        y = batch["mask"]
+        other_keys = batch.keys() - {"image", "mask", "filename"}
+        rest = {k:batch[k] for k in other_keys}
+        model_output: ModelOutput = self(x, **rest)
+        loss = self.val_loss_handler.compute_loss(model_output, y, self.criterion, self.aux_loss)
+        self.val_loss_handler.log_loss(self.log, loss_dict=loss, batch_size=x.shape[0])
```
[the remark this answers is not in the document]
Comment (Joao-L-S-Almeida) [reply terratorch/tasks/regression_tasks.py]: Totally removed. Scalar regression was considered unnecessary.

File: terratorch/tasks/regression_tasks.py [hunk 7]
```diff
@@ -185,22 +185,45 @@ def __init__(
...
     # overwrite early stopping
     def configure_callbacks(self) -> list[Callback]:
         return []
 
     def configure_models(self) -> None:
-        self.model: Model = self.model_factory.build_model(
-            "regression", aux_decoders=self.aux_heads, **self.hparams["model_args"]
-        )
+
+        self.model: Model = self.model_builder()
```
[the remark this answers is not in the document]
Comment (Joao-L-S-Almeida) [reply terratorch/tasks/regression_tasks.py:225]: I like the way in which we can define a model builder function and construct the model according to each case.

**Comment by romeokienzler** [comment]
this goes to 99.8

**Pull request closed without merging**